zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -611,48 +590,61 @@ private[consumer] final class Runloop(
       _ => command.succeed.as(state.copy(subscription = command.subscription))
     )
 
-  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Unit]] =
-    ZStream
-      .mergeAll(3, 1)(
-        ZStream.repeatWithSchedule(Command.Poll, Schedule.fixed(pollFrequency)),
-        ZStream.fromQueue(requestQueue).mapChunks(c => Chunk.single(Command.Requests(c))),
-        ZStream.fromQueue(commandQueue)
-      )
-      .tap(cmd => diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd)))
-      .runFoldZIO(State.initial) { (state, cmd) =>
-        ZIO.ifZIO(isShutdown)(onTrue = handleShutdown(state, cmd), onFalse = handleOperational(state, cmd))
-      }
+  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Any]] =
+    pollLoop(false, State.initial)
       .tapErrorCause(cause => ZIO.logErrorCause("Error in Runloop", cause))
       .onError(cause => partitions.offer(Take.failCause(cause)))
-      .unit
       .forkScoped
+
+  private def pollLoop(initialized: Boolean, state: State): ZIO[Any, Throwable, State] = {
+    def processCommands(state: State) = for {
+      // Gather available commands or return immediately if nothing in the queue
+      commands   <- commandQueue.takeAll
+      isShutdown <- isShutdown
+      updatedState <- ZIO.foldLeft(commands)(state) { case (s, cmd) =>
+                        (if (isShutdown) handleShutdown(s, cmd) else handleOperational(s, cmd)) <*
+                          diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd))
+                      }
+    } yield updatedState
+
+    def doPollIfPendingActions(state: State) =
+      for {
+        isRebalancing <- isRebalancing
+        isInitialized = initialized || state.pendingRequests.nonEmpty
+        shouldPoll =
+          state.isSubscribed && (state.pendingRequests.nonEmpty || state.pendingCommits.nonEmpty || !isInitialized || isRebalancing)
```

Comment:
`state.pendingRequests.nonEmpty` is used twice, once in a positive way and once in a negative way. In other words: `shouldPoll` is equal to `state.isSubscribed` :)

```suggestion
        shouldPoll =
          state.isSubscribed && (state.pendingRequests.nonEmpty || state.pendingCommits.nonEmpty || !initialized || isRebalancing)
```

I did not look up what `initialized` does yet. However, from the name I would guess we need:
```suggestion
        shouldPoll =
          state.isSubscribed && initialized && (state.pendingRequests.nonEmpty || state.pendingCommits.nonEmpty || isRebalancing)
```